### PR TITLE
perf(mito): Use a heap to merge batches for the same key

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?
This PR uses a heap to merge sorted batches in the `BatchMerger` instead of concatenating these batches and uses `sort_and_dedup()`. It also avoids concatenating batches as it introduces additional overhead.

These changes can speed up the merge reader when input batches don't have too many overlapping rows.

On Apple M1 Pro, merging 148797 sorted batches with 100,000,000+ rows took 2.2s before and 358ms now. Note that this only considers the cost of `BatchMerger`, not the whole `MergeReader`.

|  Before  | After |
|  ----  | ----  |
| 2.2s | 358ms |

It also fixes an issue that if a batch is empty after `filter_deleted()`, the reader will return `None` even if there are remaining batches.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/mito2/src/read/merge.rs [hunk 1]
```diff
@@ -201,32 +219,143 @@ impl BatchMerger {
...
 
-        let batches = mem::take(&mut self.batches);
-        // Concat all batches.
-        let mut batch = Batch::concat(batches)?;
-
-        // TODO(yingwen): metrics for sorted and unsorted batches.
-        if !self.is_sorted {
-            // Slow path. We need to merge overlapping batches. For simplicity, we
-            // just sort the all batches and remove duplications.
-            batch.sort_and_dedup()?;
```
Comment (killme2008) [inline src/mito2/src/read/merge.rs:217]: Question: can the `sort_and_dedup` have a better performance if all batches are small?

File: src/mito2/src/read/merge.rs [hunk 2]
```diff
@@ -201,32 +219,143 @@ impl BatchMerger {
...
+            }
+
+            // Safety: Batches (top, next) in the heap is not empty, so we can use unwrap here.
+            // Min timestamp in the next batch.
+            let next_min_ts = next.first_timestamp().unwrap();
+            let timestamps = top.timestamps_native().unwrap();
+            // Binary searches the timestamp in the top batch.
+            // Safety: Batches should have the same timestamp resolution so we can compare the native
+            // value directly.
+            match timestamps.binary_search(&next_min_ts.value()) {
```
Comment (killme2008) [inline src/mito2/src/read/merge.rs:267]: So the top batch is always sorted and deduped, am I right? Or is every batch already sorted and deduped but must be merged, deduped, and sorted?

File: src/mito2/src/read/merge.rs [hunk 3]
```diff
@@ -201,32 +219,143 @@ impl BatchMerger {
...
+
+            if top.last_timestamp() < next.first_timestamp() {
+                // If the top-most batch doesn't overlaps with the next batch, we can output it.
+                output_batch(&mut output, top)?;
+                continue;
+            }
+
+            // Safety: Batches (top, next) in the heap is not empty, so we can use unwrap here.
+            // Min timestamp in the next batch.
+            let next_min_ts = next.first_timestamp().unwrap();
```
Comment (killme2008) [inline src/mito2/src/read/merge.rs:262]: Question: Is it possible that the next batch contains more than one timestamp that is equal to `next_min_ts`?  If so, skipping just the `first_timetamp` is not enough.

**Review by killme2008 — COMMENTED** [review]
Great work! I have some questions that are already in comments.

File: src/mito2/src/read/merge.rs (hunk 2 above)
Comment (evenyag) [reply src/mito2/src/read/merge.rs:267]: Every batch is already sorted and deduped.

File: src/mito2/src/read/merge.rs (hunk 3 above)
Comment (evenyag) [reply src/mito2/src/read/merge.rs:262]: A batch itself is deduped.

File: src/mito2/src/read/merge.rs (hunk 1 above)
Comment (evenyag) [reply src/mito2/src/read/merge.rs:217]: Maybe. Concatenating and sorting batches are costly so I guess `sort_and_dedup()` might outperform merging by heap only when the total number of rows is small enough. However, we need some benchmarks to prove this and the cost between the two solutions in this situation should be close. We can try this optimization later if this is really a common case.

**Review by killme2008 — APPROVED** [review]
LGTM

**Review by fengjiachun — APPROVED** [review]
LGTM

**Pull request merged**